
enum PacketType {
  Invalid = 0;
  BeginHandshake = 1;
  DenyHandshake = 2;
  AcceptHandshake = 3;
  CompleteHandshake = 4;
  Ping = 5;
  Pong = 6;
  Close = 7;
  RequestPeers = 8;
  ListPeers = 9;
  RequestShards = 10;
  ListShards = 11;
}

message IUser {
  required string id = 1;
  required string pubkey = 2;
  required string nickname = 3;
  optional string first_authed = 4;
  optional string last_authed = 5;
}

message IPeer {
  required string ip = 1;
  required uint32 port = 2;
  required IUser user = 3;
}

message IShard {
  required string id = 1;
  required string name = 2;
  required string desc = 3;
  required bool public = 4;
  required bytes meta = 5;
  repeated IPeer peers = 6;
}

message Packet {
  required PacketType type = 1;
  required bytes data = 2;
}

message PacketBeginHandshake {
  required IPeer peer = 1;
  required uint32 timestamp = 3;
  required uint32 challenge = 4;
}

message PacketDenyHandshake {
  required string reason = 1;
  required uint32 backoff = 2;
}

message PacketAcceptHandshake {
  required IPeer peer = 1;
  required bytes response = 3;
  required uint32 challenge = 4;
}

message PacketCompleteHandshake {
  required bytes response = 1;
}

message PacketPing {
  required uint32 timestamp = 1;
  required uint32 challenge = 2;
}

message PacketPong {
  required uint32 timestamp = 1;
  required bytes response = 2;
}

message PacketClose {
  optional string reason = 1;
}

message PacketRequestPeers {
  required uint32 maxsize = 1;
  repeated string shards = 2;
}

message PacketListPeers {
  repeated IPeer peers = 1;
}

message PacketRequestShards {
  required uint32 maxsize = 1;
  repeated string shards = 2;
  optional bool peers = 3;
}

message PacketListShards {
  repeated IShard shards = 1;
}

